src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+    int position = isUU ? (size - 1) * 2 : size - 1;
+    if (isUU) {
+      __ vpbroadcastw(XMM0, Address(needle, position), Assembler::AVX_256bit);
+    } else if (isUL) {
+      // Expand needle
+      __ movzbl(rax, Address(needle, position));
+      __ movdl(XMM0, rax);
+      // Byte of needle to words
+      __ vpbroadcastw(XMM0, XMM0, Assembler::AVX_256bit);
+    } else {
+      __ vpbroadcastb(XMM0, Address(needle, position), Assembler::AVX_256bit);
+    }
+
+    // Compare next byte.  Keep the comparison mask in mask, which will
+    // accumulate
+    if (isU) {
+      __ vpcmpeqw(XMM1, XMM0, Address(haystack, position), Assembler::AVX_256bit);
+    } else {
+      __ vpcmpeqb(XMM1, XMM0, Address(haystack, position), Assembler::AVX_256bit);
+    }
+    __ vpmovmskb(tmp, XMM1, Assembler::AVX_256bit);
+    __ andq(mask, tmp);  // Accumulate matched bytes
+    __ testl(mask, mask);
+    __ je(L_noMatch);
+
+    if (size != OPT_NEEDLE_SIZE_MAX) {
+      // Found a match for this needle size
+      __ cmpq(needle_len, size);
+      __ je(L_foundall);
+    }
+  }
+
+  __ bind(L_foundall);
+  __ tzcntl(rax, mask);
+
+  if (isU) {
+    __ shrq(rax, 1);
+  }
+
+  __ bind(L_out);
+  __ addptr(rsp, COPIED_HAYSTACK_STACK_SIZE);
+  __ vzeroupper();
+  __ leave();
+  __ ret(0);
+
+  __ bind(L_noMatch);
+  __ movq(rax, -1);
+  __ jmpb(L_out);
+}
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //
+  // Set up jump table entries for both small and large haystack switches.
+
+static void setup_jump_tables(StrIntrinsicNode::ArgEncoding ae, Label &L_error, Label &L_checkRange, Label &L_fixup,
+                              address *jump_table, address *jump_table_1, MacroAssembler *_masm) {
+  bool isLL = (ae == StrIntrinsicNode::LL);
+  bool isUL = (ae == StrIntrinsicNode::UL);
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+
+  address large_hs_jmp_table[NUMBER_OF_CASES];  // Jump table for large haystacks
+  address small_hs_jmp_table[NUMBER_OF_CASES];  // Jump table for small haystacks
+  int jmp_ndx = 0;
+
+  {
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //
+    // Small haystack (<32 bytes) switch
+
+    ////////////////////////////////////////////////
+    //  On entry to each case of small_hs, the register state is:
+    //
+    //  rax = unused
+    //  rbx = &haystack
+    //  rcx = haystack length
+    //  rdx = &needle
+    //  rsi = haystack length
+    //  rdi = &haystack
+    //  rbp = -1
+    //  r8  = unused
+    //  r9  = unused
+    //  r10 = hs_len - needle len
+    //  r11 = unused
+    //  r12 = needle length
+    //  r13 = (needle length - 1)
+    //  r14 = &needle
+    //  r15 = unused
```

Comment:
There is quite a bit of redundancy in register usage. Its not incorrect, but looks odd. Not clear if this duplication can easily be removed (or if/why needed). 
```
    //  rbx = &haystack
    //  rdi = &haystack
    //  rdx = &needle
    //  r14 = &needle
    //  rcx = haystack length
    //  rsi = haystack length
    //  r12 = needle length
    //  r13 = (needle length - 1)
    //  r10 = hs_len - needle len
    //  rbp = -1
    
    //  rax = unused
    //  r11 = unused
    //  r8  = unused
    //  r9  = unused
    //  r15 = unused
```

(Could this comment be out-of-sync with the code? Looks like only rbx, r14 and temps out of unused registers are used few lines down)

Reply:
This comment provides the full register state upon entry to each of the cases of the switch.  The duplication is an artifact of the decisions made in setup code (like checking ranges, etc.).  Each case can depend on the values of the registers to be as documented on entry.  It can use either rcx or rsi to get the haystack length, for example.